File: 47.py
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if not nums: return []
        res = [[]]
        for n in nums:
            length = len(res[-1])
            new_res = []
            for item in res:
                for j in range(length,-1,-1):
                    if j<length and item[j] == n:
                        break
                    new_res.append(item[:j] + [n] + item[j:])
            res = new_res
        return res
    

File: 47.cpp
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        std::sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> tmp;
        vector<int> visited(nums.size(), 0);
        dfs(nums, res, tmp, visited);
        return res;
    }
    void dfs(vector<int>& nums, vector<vector<int>>& res, vector<int>& tmp, vector<int>& visited){
        if(nums.size() == tmp.size()){
            res.push_back(tmp);
            return;
        }
        for(int i = 0; i < nums.size(); i++){
            if (visited[i] == 1){
                continue;
            }
            if (i > 0 and nums[i] == nums[i-1] and visited[i-1] == 0){
                continue;
            }
            visited[i] = 1;
            tmp.push_back(nums[i]);
            dfs(nums, res, tmp, visited);
            visited[i] = 0;
            tmp.pop_back();
        }
        return;
    }
    
   
   
};
